tsl/src/chunk.c

```diff
@@ -259,3 +260,92 @@ chunk_invoke_drop_chunks(Oid relid, Datum older_than, Datum older_than_type)
 
 	return num_results;
 }
+
+/*
+ * drop_chunk_replica:
+ *
+ * This function drops a chunk on a specified data node. It then
+ * removes the metadata about the association of the chunk to this
+ * datanode on the access node.
+ */
+Datum
+drop_chunk_replica(PG_FUNCTION_ARGS)
+{
+	const char *node_name = PG_ARGISNULL(0) ? NULL : NameStr(*PG_GETARG_NAME(0));
+	Oid chunk_relid = PG_ARGISNULL(1) ? InvalidOid : PG_GETARG_OID(1);
+	List *data_nodes, *replicas;
+	ForeignServer *server;
+	Chunk *chunk;
+	ChunkDataNode *cdn;
+	char *drop_cmd;
+
+	if (!OidIsValid(chunk_relid))
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("Invalid chunk specification %u", chunk_relid)));
+
+	chunk = ts_chunk_get_by_relid(chunk_relid, false);
+
+	if (NULL == chunk)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE), errmsg("Invalid chunk specified")));
+
+	/* It has to be a foreign table chunk */
+	if (chunk->relkind != RELKIND_FOREIGN_TABLE)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("\"%s\" is not a valid remote chunk", get_rel_name(chunk_relid))));
+
+	server = data_node_get_foreign_server(node_name, ACL_USAGE, true, false);
+	Assert(NULL != server);
+
+	/* Early abort on missing permissions */
+	ts_hypertable_permissions_check(chunk_relid, GetUserId());
+
+	/* check that the chunk is indeed present on the specified data node */
+	cdn = ts_chunk_data_node_scan_by_chunk_id_and_node_name(chunk->fd.id,
+															server->servername,
+															CurrentMemoryContext);
+
+	if (!cdn)
+		ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_OBJECT),
+				 errmsg("Chunk \"%s\" does not exist on data node \"%s\"",
+						get_rel_name(chunk_relid),
+						node_name)));
+
+	/* Ensure that this chunk remains replicated so as to avoid data loss */
+	replicas = ts_chunk_data_node_scan_by_chunk_id(chunk->fd.id, CurrentMemoryContext);
+
+	/*
+	 * There should be atleast one surviving replica after the deletion here.
+	 *
+	 * We could fetch the corresponding hypertable and check its
+	 * replication_factor. But the user of this function is using it
+	 * to move chunk from one datanode to another and is well aware of
+	 * the replication_factor requirements
+	 */
+	if (list_length(replicas) <= 1)
```

Comment:
Can you also add a warning if the number of replicas will be less than the replication factor? E.g.
```c
if (list_length(replicas) < ht->fd.replication_factor)
		ereport(WARNING,
				(errcode(ERRCODE_TS_INSUFFICIENT_NUM_DATA_NODES),
				 errmsg("insufficient number of data nodes"),
				 errdetail("There are not enough data nodes to replicate chunks according to the"
						   " configured replication factor."),
```

Reply:
Please read the comment above this. We are not fetching the HT at all here. Also, this API will be used as part of `chunk move` where the user knows that temporarily replication factor might not be honored.

Reply:
I agree with @nikkhils that we probably should elide these  types of warnings. Especially considering the fact that this might be invoked on many chunks as part of a single "move" operation. We don't want to print one warning per chunk in that case.

Reply:
> this API will be used as part of `chunk move` where the user knows that temporarily replication factor might not be honored.

In the case of the move the replica will be already added to this moment. It is similar to dropping the last chunk replica. Using your reasoning no error should be produced on dropping the last replica, however as I noted the chunk already should be copied, so it is correct to produce the error.

Reply:
@erimatnor In the case of move the warning will be fired only if the chunk was under-replicated before calling the move.

Reply:
@k-rus but if we assume the chunk replica is moved (and not just deleted), then we aren't changing the replication state (i.e., not making it worse) compared to before the operation. So, it doesn't seem worthwhile to warn about a situation that already existed prior to the operation. Or am I missing something?

Reply:
It is worth to warn if the function is used without move. I mean it is the same as returning the error of deleting the last chunk replica. It sounds strange to me to do one but not the other, while the other places do both.

The only case when it might be confusing to give the warning is when the chunk was under-replicated before the move. May be it is good to pass a parameter to this SQL function to avoid warning?

Reply:
I believe this function will be use in two cases, either you:

1. Move the chunk (number of chunk replicas doesn't change)
2. Copy the chunk (number of chunk replicas increases).

In none of the cases will we delete the last chunk replica.

I think, in practice, the error raised when trying to delete the last chunk replica will never be raised if this function is used properly; the chunk would have been copied or moved prior to calling this function, so there should be at least one an extra replica in addition to this one. So, the previous check for last chunk replica is a pure safety check in case this is called outside the scope of the two cases above.